api/src/main/java/org/openmrs/module/rwandaemr/integration/UpdateClientRegistryPatientListener.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.openmrs.Patient;
+import org.openmrs.api.PatientService;
+import org.openmrs.module.rwandaemr.event.PatientEventListener;
+import org.openmrs.util.OpenmrsUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.jms.MapMessage;
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.Objects;
+
+/**
+ * Updates the client registry
+ */
+@Component
+public class UpdateClientRegistryPatientListener extends PatientEventListener {
+
+	protected Log log = LogFactory.getLog(getClass());
+
+	private final PatientService patientService;
+	private final IntegrationConfig integrationConfig;
+	private final ClientRegistryPatientProvider clientRegistryPatientProvider;
+
+	private static boolean processing = false;
+	private final ObjectMapper mapper = new ObjectMapper();
+	private File messagesDir;
+
+	public UpdateClientRegistryPatientListener(
+			@Autowired PatientService patientService,
+			@Autowired IntegrationConfig integrationConfig,
+			@Autowired ClientRegistryPatientProvider clientRegistryPatientProvider) {
+		this.patientService = patientService;
+		this.integrationConfig = integrationConfig;
+		this.clientRegistryPatientProvider = clientRegistryPatientProvider;
+	}
+
+	@Override
+	public void handlePatient(String patientUuid, MapMessage mapMessage) {
+		if (!integrationConfig.isHieEnabled()) {
+			log.debug("Integration with client registry is not enabled, returning");
+			return;
+		}
+		log.warn("Updating client registry with: " + patientUuid);
+		Patient patient = patientService.getPatientByUuid(patientUuid);
+		try {
+			clientRegistryPatientProvider.updatePatientInClientRegistry(patient);
+		}
+		catch (Exception e) {
+			log.warn("Error updating client registry, adding to queue: " + patientUuid + "; " + e.getMessage());
+			addPatientToQueue(patientUuid, mapMessage);
+		}
+	}
+
+	@Override
+	public void handleException(Exception e) {
+		log.error("Unexpected exception in " + getClass(), e);
+	}
+
+	public void addPatientToQueue(String patientUuid, MapMessage mapMessage) {
+		try {
+			String action = mapMessage.getString("action");
+			if (StringUtils.isEmpty(action)) {
+				throw new IllegalArgumentException("Unable to retrieve action from MapMessage");
+			}
+			Date eventDate = new Date();
+			ClientRegistryPatientQueueItem queueItem = new ClientRegistryPatientQueueItem();
+			queueItem.setPatientUuid(patientUuid);
+			queueItem.setEventType(action);
+			queueItem.setEventDatetime(eventDate);
+			writeMessageToFile(queueItem);
+		}
+		catch (Exception e) {
+			throw new IllegalStateException("Error handling patient message", e);
+		}
+	}
+
+	public void processQueuedMessages() {
```

Comment:
Seems odd to build out all the queue processing functionality and definitions within the listener itself?  Or is there a reason behind this?

Reply:
We can evolve as needed.  I was trying to keep logical code together.